ksql-engine/src/main/java/io/confluent/ksql/planner/plan/StructuredDataSourceNode.java

```diff
@@ -255,10 +256,43 @@ private Integer findMatchingTimestampField(final String timestampFieldName) {
     return null;
   }
 
+  @SuppressWarnings("unchecked")
+  private KStream<?, GenericRow> createKStream(
+      final StreamsBuilder builder,
+      final TimestampExtractor timestampExtractor,
+      final Serde<GenericRow> genericRowSerde) {
+    final KsqlStream ksqlTable = (KsqlStream) getStructuredDataSource();
+
+    if (ksqlTable.isWindowed()) {
```

Comment:
Only `ktable` can be windowed. Here seems that you are assuming `kstream` can also be windowed!

Reply:
The code does allow streams to be windowed. 

You can create a stream from a windowed table by just doing a `CREATE STREAM x AS SELECT * FROM WindowedTable`.  

You can also import a topic that contains windowed table data as a stream with `CREATE STREAM x WITH(... window_type='session')`

Reply:
You cannot create stream from a windowed table. Here is the message you will see if you try the example statement you provided:
```
ksql> create stream foo as select * from WindowedTable;
Invalid result type. Your SELECT query produces a TABLE. Please use CREATE TABLE AS SELECT statement instead.
```

Reply:
Why can't I create a stream from a table?  The stream should just be the CDC values being persisted to the underlying topic...

Reply:
It comes back to the way we interpret the data. If the data in a topic is CDC (aka, Change Data Capture) it means that the messages are not independent of each other and you should interpret the data as Table.
Of course you can interpret the data as stream if you need to, but if your key is windowed, it does not have the common `WindowedKey` interpretation and it is only used for deserialization of the key.

Reply:
Not sure I follow...

> ..., but if your key is windowed, it does not have the common WindowedKey interpretation and it is only used for deserialization of the key.

What does this mean?

Reply:
To clarify, my objection is that `isWindowed` is a `KTable` property and should only be in `SchemaKTable` class and not in `SchemaKStream` class.

Reply:
I'm probably missing something, but it doesn't really make sense to me that we expose the window type of a source to the user. At the moment the fact that a source is windowed has no bearing on the semantics of a query that consumes that source. Its just additional metadata encoded in the record key. So if it doesn't matter that a source is windowed, why add a property to specify that it is?

We can't really un-expose this for tables because its a property of the table itself that I get back when I describe the table. It also maybe makes sense for tables because KafkaStreams has a notion of a windowed table (though I don't see how it impacts the semantics of operations on it). But I don't think it makes sense to add such a property for streams.

cc @mjsax

Reply:
As per the example above, I can import a topic containing windowed data, i.e. the change log topic behind a windowed table or the output from a KStreams app containing windowed data.  

The keys in this topic will have the window data stored in them, (either start time for hopping/tumbling, or both start and end time for session windows).

If imported without telling KSQL the data, i.e. the key, is windowed, then it won't deserialize properly, (as we assume keys are strings).  

Therefore, (as per the test in `session-windows.json`), you can import windowed data into KSQL using:

```
CREATE STREAM STREAM_OF_WINDOWED_TABLE_CDC (...) WITH ( WINDOW_TYPE='SESSION', ...);
```

Causing KSQL to correctly deserialize the key.

Some of this was already supported ... ish, using the `WINDOWED` property. Though this assumed all windowed data was using a `TimeWindow`, which only serialises the start time to the key.  Hence switching this to something that supports different window types.

Hence... it seems to me that a stream can contain windowed data. 

Now, I'm not saying that KSQL can handle a windowed stream in all flavours of queries, but in the ones I've tried it seems to work. If you've got specific examples of stuff that doesn't work, then I can take a look.

We may also just be in disagreement on terminology - just because we have a stream on data where the  data/key is windowed, does that make the stream windowed in a KStreams / KSQL sense? Now that I'm not sure of. cc @mjsax @hjafarpour ?

Reply:
I've raised PR #2071 to hopefully address the issues here. Let's move the discussion there.